### ISSUE TITLE
how to test an addition to 01-main?

### ISSUE THREAD

**Issue by peutch** [wimpysworld/deb-get#714]
Hi,

I cloned the git but then running for instance:
```
$ ./deb-get list
```
in the local directory does not handle the added package into its local 01-main.
Is there an easy way to do such a test?

Regards,
Patrice

**Comment by philclifford** [comment]
Hi Patrice,

It depends what you want to test.  The easy way to test `your-addition` works is to copy the package definition to `/etc/deb-get/99-local.d/` and then `deb-get show your-addition` and if that seems reasonable then `deb-get install your-addition`.  
Then if you have or create a fork on github and want to you could, say, add your addition to a branch of your fork then edit `/etc/deb-get/01-main.repo` (or add an extra repo with just your test addition - that's basically the point).  This will test that you have the package definition in the right place and have correctly added it to the `manifest`.

**Comment by peutch** [comment]
Hi Phil,

Many thanks for your help!
I'm still having trouble testing `agena`. Do I also need to package and install `deb-get` to be able to run/test? Is it not possible as a simple (non-root) user?
Otherwise currently I am doing and getting this:
```
deb-get (main) $ ./deb-get show agena
  [+] Including local package agena
  [!] Please consider contributing back new entries, an issue (or raise a PR) directly at https://github.com/wimpysworld/deb-get/pulls
Agena
  Package:	agena
  Repository:	99-local
  Updater:	deb-get
  Installed:	No
  Published:	
  Architecture:	amd64 i386
  Download:	
  Website:	https://agena.sourceforge.net/
  Summary:	Agena is an easy-to-learn procedural programming language designed to be used in science, scripting, and many other applications

deb-get (main) $ sudo ./deb-get install agena
  [+] Including local package agena
  [!] Please consider contributing back new entries, an issue (or raise a PR) directly at https://github.com/wimpysworld/deb-get/pulls
  [+] Updating /var/cache/deb-get/agena.html
  [*] WARNING! Updating /var/cache/deb-get/agena.html failed. Deleting it.
/var/cache/deb-get/: est un dossier
  [!] ERROR! Failed to download . Deleting /var/cache/deb-get/...
```

Does show indicate that the `<IN-MIDDLE>` (to retrieve `$URL` and `$VERSION_PUBLISHED` from `$CACHE_FILE`) part of `agena` is functional?
```
DEFVER=1
ARCHS_SUPPORTED="amd64 i386"
get_website "https://agena.sourceforge.net/download.html"
if [ "${ACTION}" != "prettylist" ]; then
    <IN-MIDDLE>
fi
PRETTY_NAME="Agena"
WEBSITE="https://agena.sourceforge.net/"
SUMMARY="Agena is an easy-to-learn procedural programming language designed to be used in science, scripting, and many other applications"
```

**Comment by peutch** [comment]
Something similar to `apt download <pkg>`

**Comment by philclifford** [comment]
Don't sudo deb-get, it handles privilege escalation where needed. You may need to create/recreate the cache directory mentioned (maybe a bug). On phone aft so can only help from memory just now. I think it should not be necessary to install deb-get, but it should self-install and that might help. It never uses the local 01-main repo directly and local edits to /etc/...01-main will get overwritten on an update.

**Comment by peutch** [comment]
I got also the point for the 2 last messages:
```
/var/cache/deb-get/: est un dossier
  [!] ERROR! Failed to download . Deleting /var/cache/deb-get/...
```

It is due to an empty value for `${FILE}` in a call to `download_deb` at the lines 225-226:
```
   225	        fancy_message error "Failed to download ${URL}. Deleting ${CACHE_DIR}/${FILE}..."
   226	        ${ELEVATE} rm "${CACHE_DIR}/${FILE}" 2>/dev/null
```
Maybe each call to `download_deb` have to be protected from this or a check inside it.

Of course I understand that this only happens because of my test trying to define `agena`.

**Comment by philclifford** [comment]
I suspect your "<IN-MIDDLE>" needs tweaking a little.

your `deb-get show agena` must report a value for Download ($URL) and Published ($VERSION_PUBLISHED) .  Your $CACHE_FILE has probably downloaded OK (mine has ;-) ) so you can tweak the grepping manually on that file : 

```shell
# some variable setting  leads to ..
grep -o "http://downloads.sourceforge.net/${APP}/.*${HOST_ARCH}\.deb" "${CACHE_FILE}"
http://downloads.sourceforge.net/agena/agena-2.35.3-linux.i386.deb
http://downloads.sourceforge.net/agena/agena-2.35.3-linux.amd64.deb
http://downloads.sourceforge.net/agena/agena-2.34.5-raspi.stretch.armv6-32.deb
http://downloads.sourceforge.net/agena/agena-2.34.5-raspi.stretch.armv8-64.deb
$ HOST_ARCH=$(dpkg --print-architecture)
$ grep -o "http://downloads.sourceforge.net/${APP}/.*${HOST_ARCH}\.deb" "${CACHE_FILE}"
http://downloads.sourceforge.net/agena/agena-2.35.3-linux.amd64.deb

# which then get us to ..
$ deb-get show agena
Agena
  Package:	agena
  Repository:	99-local
  Updater:	deb-get
  Installed:	No
  Published:	2.35.3
  Architecture:	amd64 i386
  Download:	http://netcologne.dl.sourceforge.net/project/agena/Binaries/Agena%202.35.3/agena-2.35.3-linux.amd64.deb
  Website:	https://agena.sourceforge.net/
  Summary:	Agena is an easy-to-learn procedural programming language designed to be used in science, scripting, and many other applications
```

I got that with this - (worth tweaking further to maybe add support for the Raspberry Pi debs)

```shell
DEFVER=1
ARCHS_SUPPORTED="amd64 i386
#armv6-32 armv8-64"
get_website "https://agena.sourceforge.net/download.html"
if [ "${ACTION}" != "prettylist" ]; then
    local REDIR_URL="$(grep -o "http://downloads.sourceforge.net/${APP}/.*${HOST_ARCH}\.deb" "${CACHE_FILE}")"
    URL=$(unroll_url "${REDIR_URL}")
    VERSION_PUBLISHED="$(echo "${URL}" | cut -d'-' -f2)"
fi
PRETTY_NAME="Agena"
WEBSITE="https://agena.sourceforge.net/"
SUMMARY="Agena is an easy-to-learn procedural programming language designed to be used in science, scripting, and many other applications"
```

Hope that helps and I look forward to your PR.

Yes we could add more checks so that "imperfect" definitions are identified and failed earlier and maybe with more helpful errors.  Perhaps you could open another issue about that so it can be considered or worked on.  It would help avoid  contributors and developers seeing messages like this when they get their `regex` wrong and yolo it anyway: 

``` 
[+] Updating /var/cache/deb-get/agena.html
  [!] ERROR! Failed to download text-align: center; vertical-align: middle;. Deleting /var/cache/deb-get/text-align: center; vertical-align: middle;...
```
